### PR TITLE
Port to Hercules

### LINKED ISSUE #1588: Port to Hercules
**Description**

Update the system to support the new MSU Hercules.

**Dependencies**

- [x] EPIC hpc-stack/spack-stack available
- [x] Staged fix files in place (if not accessible from Orion side) - update `link_workflow.sh` to add Hercules on Orion line
- [x] Staged ICs in place (if not accessible from Orion side) - update `config.coupled_ic` for Hercules (same as Orion)
- [x] Test GDA in place (if not accessible from Orion side) - add Orion `DMPDIR` to Hercules hosts file
- [x] Workflow updates (e.g. env)
- [x] ufs-weather-model support: https://github.com/ufs-community/ufs-weather-model/pull/1733
- [x] UPP support https://github.com/NOAA-EMC/UPP/pull/760
- [x] UFS_UTILS support https://github.com/ufs-community/UFS_UTILS/pull/880
- [x] gfs-utils support
- [x] GSI-EnKF support https://github.com/NOAA-EMC/GSI/issues/574
- [x] GSI-Utils support
- [ ] GSI-Monitor support
- [ ] GDASApp support https://github.com/NOAA-EMC/GDASApp/pull/774
- [ ] EMC_verif-global support
- [x] OBSPROC packages support (obsproc/prepobs) https://github.com/NOAA-EMC/prepobs/issues/21
- [x] TC_tracker support https://github.com/NOAA-EMC/TC_tracker/issues/3
- [x] Fit2Obs support ~https://github.com/NOAA-EMC/Fit2Obs/issues/15~ Not necessary to port.  The Orion executables work fine on Hercules.

spack-stack is available on hercules via:
```sh
module purge
module use /work/noaa/epic-ps/role-epic-ps/spack-stack/modulefiles
```

Information from admin announcement:

A new computing resource is now available for the NOAA R&D HPC user community at Mississippi State University (MSU).
The new system has been named “Hercules” and as with Orion, is owned and managed by MSU. As this is a brand new
system, with a brand new software stack, we would like your help in flushing out any issues before we place the system
into full production. So we are asking our current Orion users to assist us with the pre-operational testing of this new
system. We would greatly appreciate it if you could try compiling your models, running your models, testing your
workflows, and then provide us with feedback by emailing “rdhpcs.orion.help@noaa.gov”. Please also email any questions
or issues to the same email address.  If everything goes well with the testing then we hope to announce full production
in early May. Thank you for all your help!

Hercules System Overview:

Manufacturer: Dell EMC
Model: PowerEdge C6520
Interconnect: Mellanox Infiniband NDR-200
Processor: Xeon Platinum 8380 40Core@2.3GHz
Total System Memory: 262,144 GB
Total Compute Nodes: 512
Cores per Node: 80
Total Cores: 40,960
File Systems: 2 DDN Lustre File system /work & /work2 (shared with Orion)
Allocations: Core-hour allocations (independent from Orion), Disk allocations (Shared between Orion and Hercules)
Home File System: NFS with 10GB of space per user
Other Node Types: Login nodes (4), Development nodes (2), and Data Transfer nodes (4)

MSU’s Official Hercules Documentation:

https://intranet.hpc.msstate.edu/helpdesk/resource-docs/hercules_guide.php

NOAA RDHPCS Supplemental Documentation

How to run jobs:

https://oriondocs.rdhpcs.noaa.gov/wiki/index.php/Running_Jobs_on_MSU-HPC

Please note the following:

During the pre-operational test phase we are only allowing “windfall” QOS jobs to run. This will allow you to run test jobs
without negatively impacting your project’s Fairshare. Once the system is ready for production then we will upload the
core-hour allocations and make all QOSs available for use.

Known Issues (supplemental):

https://oriondocs.rdhpcs.noaa.gov/wiki/index.php/Known_Issues

Differences between Orion and Hercules:

https://oriondocs.rdhpcs.noaa.gov/wiki/index.php/Frequently_Asked_Questions#What_are_the_differences_between_Orion_and_Hercules.3F

### PR DESCRIPTION
**Pull request by DavidHuber-NOAA**
# Description
Adds cycled support for Hercules (excluding gsi-monitor).

Partially resolves #1588

GSI monitoring is disabled on Hercules due to missing Perl modules.  That will be enabled in a later PR.

Depends on noaa-emc/gsi-utils#23
# Type of change

- New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? YES/NO

# How has this been tested?
C96/C48 tests on Hercules and Hera, C384/C192 test on Hercules

# Checklist
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes

### CODE REVIEW

**Comment by DavidHuber-NOAA** [comment]
~GEFS failed on Hercules, which is not surprising as I did not port the GEFS components.  I will save that for a separate PR and disable the GEFS CI tests on Hercules for the short term.~

Disregard.  This isn't true.  As @TerrenceMcGuinness-NOAA pointed out, this is the ln error.

Yes just confirmed, there is no such REPO **ioda-converters** in **JCSDA-interal** on **GitHub**.
```
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ pwd
/work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/gdas.cd
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ grep ioda-converter CMakeLists.txt 
    ecbuild_bundle( PROJECT iodaconv GIT "https://github.com/JCSDA-internal/ioda-converters.git" TAG c4f1731 )
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ 
```

**Comment by DavidHuber-NOAA** [comment]
@TerrenceMcGuinness-NOAA Can you point me to the build directory on Orion?  I'd like to take a look at the failure.

Sure ... Here is the Hercules fail first (You can see there is a ioda-data REPO but no such ioda-convers REPO)

```
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ pwd
/work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/gdas.cd
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ grep ioda-converter CMakeLists.txt 
    ecbuild_bundle( PROJECT iodaconv GIT "https://github.com/JCSDA-internal/ioda-converters.git" TAG c4f1731 )
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ git ls-remote -h  https://github.com/JCSDA-internal/ioda-data.git | head
3f9b5db49f3d39d296850a752de932fa87df6546	refs/heads/bugfix/IodaV2Failures
c63b9fc0d18eac869c221a865a1be384a1042031	refs/heads/bugfix/closeFilesSinglePool
fe1b77edade8e675e728e2b341bae153290f136d	refs/heads/bugfix/ioda623
071c65b5e10d361b4b7b8d2151410d58002967d9	refs/heads/develop
45a537abdc890d8e849f69a03e18c8eff7db0a95	refs/heads/feature/4max
1e74c59375593ee82ead7457e76933617040f32f	refs/heads/feature/SondeNewVariables
a609241c4ae9bbc67b9c7ef7763d10200002ef50	refs/heads/feature/SplitYamls
5a619e36e4f71f23c33dcfb8ef8afe630c2ceb1f	refs/heads/feature/TimeWindowGPSRO
c997ff30abeba7e94bd8ee0c33502d4551cecc16	refs/heads/feature/WriteTemplates
67c1228bdb6ab26d8721ecbdf06e176004e68136	refs/heads/feature/accumulators
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ git ls-remote -h  https://github.com/JCSDA-internal/ioda-converters.git
Username for 'https://github.com': ^C
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ git ls-remote -h  https://github.com/JCSDA-internal/ioda-foofighter.git
Username for 'https://github.com': 
```

**Comment by TerrenceMcGuinness-NOAA** [comment]
Ok @DavidHuber-NOAA  here is the accrual build error and location on Orion ...

```
mterry (Orion-login-2) logs (feature/hercules) $ pwd
/work2/noaa/stmp/GFS_CI_ROOT/ORION/PR/2196/global-workflow/sorc/logs
mterry (Orion-login-2) logs (feature/hercules) $ tail -20 build_gdas.log
CMake Error at /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/unified-env/install/intel/2022.0.2/ecbuild-3.7.2-ee35mw5/share/ecbuild/cmake/ecbuild_log.cmake:190 (message):
  CRITICAL -
  /work2/noaa/stmp/GFS_CI_ROOT/ORION/PR/2196/global-workflow/sorc/gdas.cd/iodaconv
  git clone failed:

    /apps/git-2.28.0/bin/git clone https://github.com/JCSDA-internal/ioda-converters.git  /work2/noaa/stmp/GFS_CI_ROOT/ORION/PR/2196/global-workflow/sorc/gdas.cd/iodaconv -q
    fatal: could not read Username for 'https://github.com': No such device or address

  

  
Call Stack (most recent call first):
  /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/unified-env/install/intel/2022.0.2/ecbuild-3.7.2-ee35mw5/share/ecbuild/cmake/ecbuild_git.cmake:107 (ecbuild_critical)
  /work/noaa/epic/role-epic/spack-stack/orion/spack-stack-1.5.1/envs/unified-env/install/intel/2022.0.2/ecbuild-3.7.2-ee35mw5/share/ecbuild/cmake/ecbuild_bundle.cmake:157 (ecbuild_git)
  CMakeLists.txt:110 (ecbuild_bundle)


-- Configuring incomplete, errors occurred!
See also "/work2/noaa/stmp/GFS_CI_ROOT/ORION/PR/2196/global-workflow/sorc/gdas.cd/build/CMakeFiles/CMakeOutput.log".
See also "/work2/noaa/stmp/GFS_CI_ROOT/ORION/PR/2196/global-workflow/sorc/gdas.cd/build/CMakeFiles/CMakeError.log".
mterry (Orion-login-2) logs (feature/hercules) $ 
```

Looks like we just need to find out why on Orion this was built but not on Hercules:

```
mterry (hercules-login-4) gdas.cd ((HEAD detached at f44a6d5)) $ ls $PWD/CMakeLists.txt 
/work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/gdas.cd/CMakeLists.txt


# Build IODA converters
  option(BUILD_IODA_CONVERTERS "Build IODA Converters" ON)
  if(BUILD_IODA_CONVERTERS)
    ecbuild_bundle( PROJECT iodaconv GIT "https://github.com/JCSDA-internal/ioda-converters.git" TAG c4f1731 )
  endif()
```

**Comment by DavidHuber-NOAA** [comment]
@TerrenceMcGuinness-NOAA the GDASApp isn't supported on Hercules yet.  It should have errored out when trying to build it.  Same with the GSI monitor.

**Comment by DavidHuber-NOAA** [comment]
Looks like build_gdas.sh, build_gsi.sh, build_gsi_utils.sh, and build_gsi_monitor.sh were never called.  In `/work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/log.build`:

```
Creating logs folder
Creating ../exec folder
Running "module reset". Resetting modules to system default. The following $MODULEPATH directories have been removed: None
Building upp, ufs, ufs_utils, ww3prepost, gfs_utils
Starting build_upp.sh
Starting build_ufs.sh
Starting build_ufs_utils.sh
Starting build_ww3prepost.sh
Starting build_gfs_utils.sh
build_ufs_utils.sh completed successfully!
build_gfs_utils.sh completed successfully!
build_upp.sh completed successfully!
build_ww3prepost.sh completed successfully!
build_ufs.sh completed successfully!

 .... Build system finished ....
```

That doesn't make sense to me, though.  The script clearly has `./build_all.sh -gu`.

T.McG: Let me check that.  It might not in the cron repo that runs that CI script.

**Comment by TerrenceMcGuinness-NOAA** [comment]
@DavidHuber-NOAA wait that was it, we had FF only builds on Hercules up until now, changing that now ... (The Hercules CI updates where never merged because of the `ln` issue and are still in the cron sace)

```
# TODO until full cycle port on Hercules is complete
# we only build for FF for now on Hercules
if [[ "${MACHINE_ID}" == "hercules" ]]; then
  ./build_all.sh  >> log.build 2>&1
else
  ./build_all.sh -gu  >> log.build 2>&1
fi
```

**Comment by TerrenceMcGuinness-NOAA** [comment]
@DavidHuber-NOAA  Hercules now fails to build in the  `detect_machine.sh` script in the gdas repo when using `./build_all.sh -gu`
```
mterry (hercules-login-1) logs (feature/hercules) $ cat /work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/logs/build_gdas.log

+ build_gdas.sh[4]OPTIND=1
+ build_gdas.sh[5]getopts :j:dv option
+ build_gdas.sh[6]case "${option}" in
+ build_gdas.sh[8]export BUILD_JOBS=8
+ build_gdas.sh[8]BUILD_JOBS=8
+ build_gdas.sh[5]getopts :j:dv option
+ build_gdas.sh[20]shift 2
+ build_gdas.sh[24]BUILD_TYPE=Release
+ build_gdas.sh[24]BUILD_VERBOSE=NO
+ build_gdas.sh[24]BUILD_JOBS=8
+ build_gdas.sh[24]WORKFLOW_BUILD=ON
+ build_gdas.sh[24]./gdas.cd/build.sh
/work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/gdas.cd/ush/detect_machine.sh: line 40: MACHINE_ID: unbound variable
```

PS: I hand changed _CI-Hera-Failed_ because its build actually failed because of the issue of not having **ioda-converters** repo in **JCSDA-internal** on **GitHub**

**Comment by DavidHuber-NOAA** [comment]
Yeah, that's not surprising.  GDAS hasn't been ported to Hercules yet.  PR noaa-emc/gdasapp#774 will bring in that capability.

Also, the GSI monitor hasn't been ported yet, either.  So we will need that before we can proceed with cycled CI testing on Hercules.  I will turn those tests off and ~reenable the basic build in Hercules CI tests~.  I see now that the basic build option was just in your cron @TerrenceMcGuinness-NOAA.  Could you disable that?

**Comment by DavidHuber-NOAA** [comment]
@CoryMartin-NOAA The GDASApp is failing to build on Hera and Orion because it cannot checkout ioda-converters.  Has that repository been made private?  If so, should we avoid building it or can it be made public again?

**Comment by RussTreadon-NOAA** [comment]
@DavidHuber-NOAA , GDASApp `CMakeLists.txt` clones the following _private_ JEDI repos
```
    ecbuild_bundle( PROJECT iodaconv GIT "https://github.com/JCSDA-internal/ioda-converters.git" BRANCH develop )
      ecbuild_bundle( PROJECT ioda-data GIT "https://github.com/JCSDA-internal/ioda-data.git" BRANCH develop )
      ecbuild_bundle( PROJECT ufo-data GIT "https://github.com/JCSDA-internal/ufo-data.git" BRANCH develop )
      ecbuild_bundle( PROJECT fv3-jedi-data GIT "https://github.com/JCSDA-internal/fv3-jedi-data.git" BRANCH develop )
```
We (EMC) do not control which repositories JCSDA makes private or public.  It may be possible for us to switch to the public snapshots of these repos hosted at https://github.com/JCSDA/

**Comment by CoryMartin-NOAA** [comment]
@DavidHuber-NOAA thanks for letting me know, I didn't know they made this private for some reason... I will escalate this to JCSDA. Stay tuned.

**Comment by WalterKolczynski-NOAA** [comment]
JCSDA-internal/ioda-data and JCSDA-internal/fv3-jedi-data both appear to be public now. JCSDA-internal/ufo-data still is not, but not sure if we need that.

**Comment by DavidHuber-NOAA** [comment]
Great, thanks @WalterKolczynski-NOAA!  I will give it a test build on Orion.

**Comment by DavidHuber-NOAA** [comment]
@WalterKolczynski-NOAA The build is still failing for me when attempting to clone ioda-converters.  Do you perhaps have read access to the repo?  It looks like the fix update (pr #2198) is failing the same way.

**Comment by WalterKolczynski-NOAA** [comment]
Looks like some needed ones are still private (@TerrenceMcGuinness-NOAA hits JCSDA-internal/ioda-converters.git)

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA The build is still failing for me when attempting to clone ioda-converters. Do you perhaps have read access to the repo? It looks like the fix update (pr #2198) is failing the same way.

I personally have access, which is why me testing does no good. ioda-converters and ufo-data are both still private.

**Comment by CoryMartin-NOAA** [comment]
Try it now, I think they have been reverted to public access again

**Comment by DavidHuber-NOAA** [comment]
Thanks, I'm starting a test build now.

**Comment by DavidHuber-NOAA** [comment]
The build was successful on Orion.  Thanks again @CoryMartin-NOAA!

**Review by aerorahul — APPROVED** [review]
lgtm
Thanks for the work.

File: parm/config/gfs/config.resources
Comment (KateFriedman-NOAA) [inline parm/config/gfs/config.resources:917]: This line earlier (line 908):
```
export npe_node_eomg=${npe_node_eobs}
```
...needs to benefit from updating `npe_node_eobs` for S4 and Hercules C384. Didn't notice earlier when you added S4, just seeing this now. :)

**Review by KateFriedman-NOAA — APPROVED** [review]
Looks good! Left one comment to consider before merging but approve otherwise. Approve pending successful completion of CI tests. Thanks @DavidHuber-NOAA !

File: parm/config/gfs/config.resources
Comment (DavidHuber-NOAA) [reply parm/config/gfs/config.resources:917]: Good catch, thanks @KateFriedman-NOAA!  Fixed.

**Comment by DavidHuber-NOAA** [comment]
Starting CI testing.

**Comment by TerrenceMcGuinness-NOAA** [comment]
@DavidHuber-NOAA  Almost have it, but still have one last detect machine script not ready for Hercules ...
```
mterry (hercules-login-1) logs (feature/hercules) $ tail -3 /work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/logs/build_gdas.log 
+ build_gdas.sh[24]WORKFLOW_BUILD=ON
+ build_gdas.sh[24]./gdas.cd/build.sh
/work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/gdas.cd/ush/detect_machine.sh: line 40: MACHINE_ID: unbound variable
```

```
mterry (hercules-login-1) logs (feature/hercules) $ grep MACHINE_ID /work2/noaa/stmp/GFS_CI_ROOT/HERCULES/PR/2196/global-workflow/sorc/gdas.cd/ush/detect_machine.sh | awk '{print $2}' | uniq 
MACHINE_ID=wcoss2
MACHINE_ID=gaea
MACHINE_ID=hera
MACHINE_ID=s4
MACHINE_ID=jet
MACHINE_ID=orion
MACHINE_ID=cheyenne
MACHINE_ID=stampede
MACHINE_ID=expanse
MACHINE_ID=discover
```

**Comment by DavidHuber-NOAA** [comment]
@TerrenceMcGuinness-NOAA The GDASApp will not build on Hercules, nor will GSI-monitor, so we need to disable cycled tests and revert to `./build_all.sh` without `-gu` for Hercules.  I have already disabled the cycled CI tests.  Can you please disable the build on cron?  Or do I need to do that in this PR?

**Comment by DavidHuber-NOAA** [comment]
@TerrenceMcGuinness-NOAA please see my comments/question [here](https://github.com/NOAA-EMC/global-workflow/pull/2196#issuecomment-1879321883).

**Comment by DavidHuber-NOAA** [comment]
I think I found a way to run the CI tests without building the monitor or gdas app.  Trying CI testing again.

**Pull request merged**
